### PR TITLE
Fixes #8729: nan is a rational function

### LINKED ISSUE #8729: `nan` is a rational function
In `is_rational_function()`, there's a line which states that if the intersection of the expression's free symbols is an empty set, then the expression is a *constant rational function*. Thus, `nan.is_rational_function()` is `True`, and this behaviour causes bugs like #8414.

I guess that a simple check of `if self is nan` or so just before `return True` should fix that bug. However, there may be some implications that I miss, so I leave it as an issue rather than as a pull request.

### PR DESCRIPTION
**Pull request by leosartaj**
Now, 

```
>>> (nan).is_rational_function()
False
```

Fixes #8729

### CODE REVIEW

**Comment by leosartaj** [comment]
@smichr

**Comment by debugger22** [comment]
Please update your branch with master.

File: sympy/core/expr.py [hunk 1]
```diff
@@ -1,6 +1,7 @@
 from __future__ import print_function, division
 
 from .core import C
+from .singleton import S
```
Comment (debugger22) [inline sympy/core/expr.py]: Duplicate import.

**Comment by leosartaj** [comment]
Updated

**Comment by debugger22** [comment]
+1

**Comment by leosartaj** [comment]
Anything needs to be done here?

**Comment by leosartaj** [comment]
Can this be merged now? @debugger22 @jksuom

**Comment by pelegm** [comment]
I'm also +1, even if it's not complete.

**Comment by jksuom** [comment]
Perhaps one could also take care of cases like ``oo.is_rational_function()``.

**Comment by leosartaj** [comment]
Should this be the behavior

```
>>> oo.is_rational_function()
False
>>> zoo.is_rational_function()
False
```
@jksuom

**Comment by jksuom** [comment]
Yes, that is what I think. Even ``-oo`` should not be a rational function. (Sums and differences of rational functions should also be rational functions. But ``oo - oo`` and ``zoo + zoo`` are ``nan``, So if ``nan`` is not a rational function, then ``oo`` and ``zoo`` must not be that either.)

**Comment by leosartaj** [comment]
I agree.

**Comment by leosartaj** [comment]
I have made the following changes as suggested by @jksuom 

```
>>> (oo).is_rational_function()
False
>>> (-oo).is_rational_function()
False
>>> (zoo).is_rational_function()
False
```

**Comment by jksuom** [comment]
+1

**Pull request merged**